web/src/copy_and_paste.ts

```diff
@@ -170,8 +206,44 @@ export function copy_handler(): void {
     // * Otherwise, we want to copy the bodies of all messages that
     //   were partially covered by the selection.
 
+    // This is done to make the copying within math blocks smarter.
     const selection = window.getSelection();
     assert(selection !== null);
+    const original_range = selection.getRangeAt(0);
+    const anchor_element = get_parent(selection.anchorNode);
+    const focus_element = get_parent(selection.focusNode);
+    if (anchor_element) {
+        const parent = anchor_element.closest(".katex-display");
+        const is_math_block = parent !== null && parent !== selection.anchorNode;
+        if (is_math_block) {
+            const range = document.createRange();
+            range.selectNodeContents(parent);
+            selection.removeAllRanges();
+            selection.addRange(range);
+        }
+    } else if (focus_element) {
+        const parent = focus_element.closest(".katex-display");
+        const is_math_block = parent !== null && parent !== selection.focusNode;
+        if (is_math_block) {
+            const range = document.createRange();
+            range.selectNodeContents(parent);
+            selection.removeAllRanges();
+            selection.addRange(range);
+        }
+    }
+
+    // If the anchor and focus end up in different katex-displays, this isn't
+    // mean to be an inline expression, like the above.
+    if (
+        focus_element &&
+        anchor_element &&
+        focus_element?.closest(".katex-display") !== anchor_element?.closest(".katex-display")
+    ) {
+        selection.removeAllRanges();
+        assert(original_range !== undefined);
+        selection.addRange(original_range);
+    }
```

Comment:
I don't understand what you're trying to accomplish here from an algorithms standpoint. Is the intent here is to mutate the selection? I think you should put these blocks of code into a function with some readable name like `expand_katex_selection_ranges`, and clear comments explaining what it's doing.